### PR TITLE
chol(A, 'lower') and sum(A, dim)

### PR DESCRIPTION
**Pull request by newebug**
Add a chol.js to produce a lower/upper triangular matrix.

Implement sum(A, dim)

### CODE REVIEW

File: lib/function/statistics/sum.js [hunk 1]
```diff
@@ -57,16 +58,41 @@ module.exports = function (math) {
...
+      console.log(A)
+      for(var i=0; i<size[0]; i++){
+        for(var j=0; j<size[1]; j++){
+          A[0][j] += array[i][j];
+        }
+      }
+      return A;
+    }
+    else if (dim === 2) {
+      var size = math.size(array)
```
Comment (BigFav) [inline lib/function/statistics/sum.js]: Missing semi-colon. Also missing one on line 75.

File: lib/function/statistics/sum.js [hunk 2]
```diff
@@ -57,16 +58,41 @@ module.exports = function (math) {
...
+      collection.deepForEach(array, function (value) {
+        sum = (sum === undefined) ? value : math.add(sum, value);
+      });
+    }
+    // implement sum(A, dim)
+    else if (dim === 1) {
+      var size = math.size(array)
+      var A = math.zeros([1, size[1]]);
+      
+      console.log(A)
```
Comment (BigFav) [inline lib/function/statistics/sum.js]: Remnant from testing? :)

**Comment by BigFav** [comment]
Hopefully these small things are causing this too fail, and nothing too major. Can you also add test cases for these please?

**Comment by newebug** [comment]
add function/matrix/repmat.js

add chol.test.js, repmat.test.js, modify sum.js, sum.test.js

**Comment by josdejong** [comment]
Nice job! Thanks for adding the tests. 

Few remarks:
- The unit tests fail on Travis: https://travis-ci.org/josdejong/mathjs/builds/46629779
- Can you add support and unit tests for BigNumbers and Complex numbers? Should work fine when consistently using functions from `math` for all arithmetic operations (like `math.multiply`, `math.subtract`, etc).

**Comment by Wikunia** [comment]
I've hopefully tried your newest version @newebug but I found a bug.
Cholesky needs a `n x n` matrix and it must be positive definite so the elements in the diagonal matrix must be `> 0`. This matrix: `[[2,2],[2,3]]` should throw an error. 
The other thing you have to check is: Is the matrix Hermitian?

**Comment by newebug** [comment]
Yes, not all matrix can be choled, I implement the chol method according matlab convention, in deed, we would get a result including imaginary number element while chol-ing a non-positive definite maxtrix. Shall we throw an exception when it attemp to chol a non-positive definite maxtrix or return a result with imaginary number?

File: lib/function/matrix/chol.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+   * See also:
+   *
+   *    inv
+   *
+   * @param  {Array | Matrix} x Array | Matrix to factor.
+   * @param  {String} [p='upper'] 'upper' || 'lower', default is 'upper'
+   * @return {Array | Matrix} the triangular matrix
+   */
+  math.chol = function chol(x, p) {
+    var array = (x instanceof Matrix) ? x.valueOf() : x;
```
Comment (rjbaucells) [inline lib/function/matrix/chol.js]: This implementation will not have a good performance on Sparse Matrices. Calling `matrix.valueOf()` is very expensive on those types of matrices. Implementation should delegate the specifics to the actual Matrix type (Dense, CCS, CRS).

**Comment by josdejong** [comment]
@newebug for some reason I overlooked your (still) open PR, sorry. I thought there was still discussion going on here, what is the current status?

**Comment by rjbaucells** [comment]
An important note is that the Cholesky factorization implemented in this pull request will not scale for Sparse matrices. Calculating the inverse of a big matrix is prohibited expensive.

**Comment by newebug** [comment]
I'll close this pr and make some fix. Thank you!

**Comment by josdejong** [comment]
Thanks @newebug

**Pull request closed without merging**